### DISCUSSION TITLE
NoneType model value examples are not supported

### DISCUSSION THREAD

**Discussion by Kylmakalle** [fastapi/fastapi#8951] [Questions]
### Description
I'm not able to pass a `None`/ as an example for a model value created with `pydantic.Field`

Looks like parser can't find a difference between  `example=None` as a NoneType variable and "_missing example field_"

```Python
class SomeResponse(BaseModel):
       city: Optional[str] = Field(..., nullable=True, example=None)
```

### Expected openapi.json
```json
"city": {
"title": "City",
"type": "string",
"nullable": true,
"example": null
}
```

### Occurred openapi.json
```json
"city": {
"title": "City",
"type": "string",
"nullable": true
}
```

### Environment

* OS: macOS
* FastAPI Version: 0.61.1
* pydantic Version: 1.6.1

* Python version: 3.7.9

**Comment by ycd** [comment]
I did not understand the problem here, why do you need to pass `example=None` in Field.

Also when I check the schema that Pydantic creates, everything seemed fine to me.

```python

SomeResponse.schema()

Out: {
    "properties":{
        "city":{
            "example":"None",
            "nullable":True,
            "title":"City",
            "type":"string"
        }
    },
    "required":[
        "city"
    ],
    "title":"SomeResponse",
    "type":"object"}
```

Can you reproduce how did you get the output?

**Comment by Kylmakalle** [comment]
> why do you need to pass example=None in Field

That's part of api response and it's intended. 

> I check the schema that Pydantic creates

But it looks **not** fine in `/openapi.json`


> Can you reproduce how did you get the output?

It's provided in issue description
> city: Optional[str] = Field(..., nullable=True, example=None)

I also attach an example there so you can be sure it's even more buggy if I provide example value via `class Config` for schema. Fields with `None` as example are missing...
https://gist.github.com/Kylmakalle/8682416c2155b99a89961d33c0d3988b

![image](https://user-images.githubusercontent.com/24507532/96488266-60719800-1246-11eb-8f6c-0798b024fcba.png)

**Comment by ycd** [comment]
Yes I can confirm this is strange, even more strange thing is when i debug your gist,  `scheme_extra` is an empty dictionary.

![image](https://user-images.githubusercontent.com/62724709/96493840-27d5bc80-124e-11eb-8be6-493a94c88540.png)

**Comment by sommergis** [comment]
Hi all,
I ran into the same issue today. It seems like None values of fields in examples in OpenAPI schema generation are being excluded by FastAPI. The API works as expected, it's only an issue in the OpenAPI schema.

**Sample Model**

```py
class APIErrorMessage(BaseModel):
    """ Model for basic API response for an unsuccessful request. """
    content: Any = None
    error: str = ""
    detail: Optional[str] = None

    class Config:
        """ Real world example for OpenAPI documentation """
        schema_extra = {
            "example": {
                "content": None,
                "error": "Authentication Error",
                "detail": "key validity expired"
            }
        }
```

**OpenAPI Response**

```json

"example": {
    "error": "Authentication Error",
    "detail": "key validity expired"
}
```


Maybe this is the location in the code? `exclude_none=True`

https://github.com/tiangolo/fastapi/blob/d2eb4a71eef08e201082d23517d4ddc99ff258b5/fastapi/openapi/utils.py#L377

**Environment**

FastAPI Version: 0.62.0

**Comment by four43** [comment]
I tried setting that `exclude_none=False` and oh boy does everything blow up. It might need to be a little more selective in what it removes however, a heavy handed exclusion of all Nones is killing off what we need.

**Comment by four43** [comment]
@Kylmakalle @sommergis - Does my little PR fix the issues you were having? ^

**Comment by abuendia** [comment]
Hi there - is there any update on this issue / ETA for the PR merge? @four43

**Comment by four43** [comment]
My PR is open and waiting for approval, I guess I don't know what more I can do :(

**Comment by abuendia** [comment]
Hi @tiangolo - Could you please take a look at this?

**Comment by abuendia** [comment]
Hi @tiangolo, @Kludex - Sorry to nag on this! Is there anything I can do to help merge @four43's PR? I am taking dependency on the fix for a prod workflow.

**Reply by four43** [reply]
Is this really a discussion if it's a bug and it has a PR to fix it? Lol

**Comment by wombatonfire** [comment]
I suppose a similar scenario is when the model contains a field, that only allows None as a value:

```
class Connection(BaseModel):
    addr: str
    port: int
    key: None
```

In this case, an auto-generated example is:

```
"conn": {
    "addr": "string",
    "port": 0,
    "key": "Unknown Type: null"
  }
```

I am not sure why `null` is an unknown type. Is this expected?

**Reply by flashgorman-chorus** [reply]
`key: None` is not valid.  `None` is a value in Python.  `key: None` is like writing `key: "foo"` which doesn't make sense.

If you want to say the value of key _may_ be `None` then you write `key: Optional[str]` (or `Optional[int]`, etc.).

**Comment by bohea** [comment]
I made a workaround

```
class SomeResponse(BaseModel):
       city: Optional[str] = Field(None, example="#null#")
```


```
from fastapi.openapi.utils import get_openapi


def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema
    openapi_schema = get_openapi(
        title=app.title,
        version=app.version,
        openapi_version=app.openapi_version,
        description=app.description,
        terms_of_service=app.terms_of_service,
        contact=app.contact,
        license_info=app.license_info,
        routes=app.routes,
        tags=app.openapi_tags,
        servers=app.servers,
    )
    components = openapi_schema["components"]
    if 'schemas' in components:
        schemas = components["schemas"]
        for _, schema in schemas.items():
            if 'properties' in schema:
                for _, p in schema['properties'].items():
                    if 'example' in p:
                        if p['example'] == "#null#":
                            p['example'] = None
    app.openapi_schema = openapi_schema
    return app.openapi_schema


app.openapi = custom_openapi

```

**Comment by FedericoHeichou** [comment]
Same problem here, I tried with every version (0.100.1 and pydantic v2 too).  
I think at least with `dict[str, int | None]` should work...

```python
from pydantic import BaseModel, Field
from fastapi import FastAPI


class Example(BaseModel):
    broken_single: dict = Field(..., example={'ok': 1, 'null_example': None})
    broken: dict = Field(..., examples=[{'ok': 1, 'null_example': None}])
    broken_specific: dict[str, int | None] = Field(..., examples=[{'ok': 1, 'null_example': None}])


print(Example.schema_json())
print(Example.model_json_schema())

app = FastAPI()

@app.post('/')
async def broken(example: Example):
    return {'ok': 1, 'null_example': None}
```

**Comment by Dmitry-k42** [comment]
Not sure how much the problem is still actual. I just faced it today and want to share my workaround.
@Kylmakalle try to write 
```python
city: Optional[str] = Field(..., default=None, examples=[None])
```
It works perfectly in my case

**Comment by oren-galaxy** [comment]
Still haven't got a solution, especially when using `json_schema_extra` ([FastAPI docs](https://fastapi.tiangolo.com/tutorial/schema-extra-example/#__tabbed_1_1)).

e.g.

I want the `i_want_this_null` field to show null in the Swagger example but it gets filtered out:
```
class Foo(BaseModel):
    bar: dict

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "bar": {
                        "a_field": "blah",
                        "i_want_this_null": None,
                    }
                }
            ]
        }
    }
```

<img width="253" alt="Screenshot 2023-11-30 at 5 51 27 PM" src="https://github.com/tiangolo/fastapi/assets/111951817/f751692c-0441-4fc6-bc83-aa1c60e5fc34">

**Comment by erasmux** [comment]
Ran into a similar problem with FastAPI v0.108.0 and it took me some time to understand the exact issue and find a workaround (didn't check any other versions).

My real use case is to provide multiple examples in a list of nested model objects:
```python
class Bar(BaseModel):
    name: str = Field(examples=["name"])
    optional: str | None = Field(examples=[None])

class Foo(BaseModel):
    bars: list[Bar] = Field(examples=[[Bar(name="full", optional="given"),
                                       Bar(name="partial", optional=None)]])
```
Resulting in the following example in the openapi.json schemas, which is missing the `"optional": null` for the second Bar:
```javascript
    "Foo": {
        "properties": {
            "bars": {
                "examples": [
                    [
                        {
                            "name": "full",
                            "optional": "given"
                        },
                        {
                            "name": "partial"
                        }
                    ]
                ],
            }
        }
    }
```

To the best of my understanding, the problem is the line that @sommergis pointed out:
https://github.com/tiangolo/fastapi/blob/3425c834cc113060d00b8a295d3456eb306a109d/fastapi/openapi/utils.py#L530
the `exclude_none=True` will filter out any dictionary values which are None.

It is important to notice that when None is a list value it is not filtered out. So for the `Bar` model above, the line:
```python
    optional: str | None = Field(examples=[None]) 
```
Will work as intended and generate the following example in the openapi.json schemas:
```javascript
    "Bar": {
        "properties": {
            "optional": {
                "examples": [
                    null
                ]
            }
        }
    }
```
resulting in a proper example in Swagger:
```javascript
{
  "name": "name",
  "optional": null
}
```

The same holds if the example is provided through the `json_schema_extra`.

The only workaround I found is basically the one suggested by @bohea above. My version of this workaround is:

```python
def encode_none(schema):
    if isinstance(schema, dict):
        return {k: "\x00None" if v is None else encode_none(v)
                for k, v in schema.items()}
    elif isinstance(schema, list):
        return [encode_none(e) for e in schema]
    else:
        return schema

def decode_none(schema):
    if isinstance(schema, dict):
        return {k: None if v == "\x00None" else decode_none(v)
                for k, v in schema.items()}
    elif isinstance(schema, list):
        return [decode_none(e) for e in schema]
    else:
        return schema

class MyFastAPI(FastAPI):
    def openapi(self):
        if not self.openapi_schema:
            schema = super().openapi()
            components = schema.get("components")
            if components and 'schemas' in components:
                components['schemas'] = decode_none(components['schemas'])
        return self.openapi_schema
```

Then for examples which include None values, I use the `json_schema_extra` to encode None values:
```
Foo.model_config = {
    "json_schema_extra": {
        "examples": encode_none([
            [
                {
                    "name": "full",
                    "optional": "given"
                },
                {
                    "name": "partial",
                    "optional": None
                }
            ]
        ])
    }
}
```
(In my actual code, I use my own subclass of BaseModel, which has helper functions to automatically get a model_dump with values initialized from the `Field(examples=...)` defined in the model subclass, so I added the `encode_none()` call there. If there is a way to automatically tweak the generated openapi examples, meaning without the need to define the model_config "json_schema_extra" for each model with examples with None, I haven't found it. Thankfully I currently have only one such model...)

**Reply by cknv** [reply]
Ran into this problem myself, but in my case the `None` values was in a different location, so I modified the example:

```python
import typing

import fastapi


def encode_none(schema):
    if schema is None:
        return "None"
    if isinstance(schema, dict):
        return {key: encode_none(value) for key, value in schema.items()}
    if isinstance(schema, list):
        return [encode_none(each) for each in schema]
    return schema


def decode_none(schema):
    if schema == "None":
        return None
    if isinstance(schema, dict):
        return {key: decode_none(value) for key, value in schema.items()}
    if isinstance(schema, list):
        return [decode_none(each) for each in schema]
    return schema


class CustomFastAPI(fastapi.FastAPI):
    def openapi(self) -> dict[str, typing.Any]:
        if not self.openapi_schema:
            spec = super().openapi()
            self.openapi_schema = decode_none(spec)
        return self.openapi_schema or {}
```

Much the same as the original example the `encode_none` is used to encode the examples (reduced for brevity):
```python
def handle_request(
    request: typing.Annotated[
        RequestModel,
        Body(
            openapi_examples=encode_none(
                {
                    "minimal": {
                        "summary": "A minimal example",
                        "value": {
                            "a_field": None,
                        },
                    },
                },
            ),
        ),
    ],
) -> Response:
    ...
```

This code of course just recurses through _the entire_ schema, but since it is cached after the first time it should not have any impact.
